src/memray/_memray.pyx

```diff
@@ -199,37 +208,66 @@ cdef class AllocationRecord:
 
     def hybrid_stack_trace(self, max_stacks=None):
         cdef vector[unsigned char] is_entry_frame
-        cdef list python_stack
-        if max_stacks is None:
-            python_stack = self._reader.get().Py_GetStackFrameAndEntryInfo(
-                self._tuple[4], &is_entry_frame
-            )
-        else:
-            python_stack = self._reader.get().Py_GetStackFrameAndEntryInfo(
-                self._tuple[4], &is_entry_frame, max_stacks
-            )
 
-        native_stack = self.native_stack_trace(max_stacks)
+        python_stack = self._reader.get().Py_GetStackFrameAndEntryInfo(
+            self._tuple[4], &is_entry_frame
+        )
+        native_stack = self.native_stack_trace()
 
-        if not python_stack:
-            yield from native_stack
-            return
+        cdef ssize_t num_non_entry_frames = 0
+        for val in is_entry_frame:
+            if not val:
+                num_non_entry_frames += 1
+
+        hybrid_stack = [None] * (len(native_stack) + num_non_entry_frames)
+
+        cdef ssize_t pidx = len(python_stack) - 1
+        cdef ssize_t nidx = len(native_stack) - 1
+        cdef ssize_t hidx = len(hybrid_stack) - 1
+
+        cdef ssize_t to_skip = 0
+        if self._tuple[0] == self._reader.get().getMainThreadTid():
+            to_skip = self._reader.get().getSkippedFramesOnMainThread()
+
+        cdef bool first_kept_python_frame = False
+        cdef bool first
+
+        while nidx >= 0:
+            assert hidx >= 0
+            native_frame = native_stack[nidx]
+            nidx -= 1
 
-        cdef size_t python_frame_index = 0
-        cdef size_t num_python_frames = len(python_stack)
-        for native_frame in native_stack:
-            if python_frame_index >= num_python_frames:
-                break
             symbol = native_frame[0]
-            if self._is_eval_frame(symbol):
-                while python_frame_index < num_python_frames:
-                    python_frame = python_stack[python_frame_index]
-                    python_frame_index += 1
-                    yield python_frame
-                    if is_entry_frame[python_frame_index - 1]:
-                        break
+            if pidx >= 0 and self._is_eval_frame(symbol):
+                first = True
+                while pidx >= 0 and (first or not is_entry_frame[pidx]):
+                    first = False
+                    assert hidx >= 0
+                    if first_kept_python_frame:
+                        del hybrid_stack[hidx + 1:]
+                        first_kept_python_frame = False
+
+                    hybrid_stack[hidx] = python_stack[pidx]
+                    hidx -= 1
+                    pidx -= 1
+
+                    if to_skip > 0:
+                        to_skip -= 1
+                        if to_skip == 0:
+                            first_kept_python_frame = True
```

Comment:
This code is quite difficult to follow. I think we should extract some of these pieces into functions at the very least.